fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+    support = int(get_weight(store, block_root))
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    maximum_support = int(
+        get_committee_weight_between_slots(justified_state, Slot(parent_block.slot + 1), current_slot)
+    )
+    proposer_score = int(get_proposer_score(store))
+
+    """
+    Returns whether the following condition is true using only integer arithmetic
+    support / maximum_support >
+    0.5 * (1 + proposer_score / maximum_support) + confirmation_byzantine_threshold / 100
+    """
+
+    return (
+        100 * support > 
+        50 * maximum_support + 50 * proposer_score + confirmation_byzantine_threshold * maximum_support 
+    )
+```
+
+#### `is_lmd_confirmed`
+
+```python
+def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    if block_root == store.finalized_checkpoint.root:
+        return True
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return False
+        else:
+            # Check is_one_confirmed for this block and is_lmd_confirmed for the preceding chain.
+            return (
+                is_one_confirmed(store, confirmation_byzantine_threshold, block_root) and
+                is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
+            )
+```
+
+#### `get_total_active_balance_for_block_root`
+
+```python
+def get_total_active_balance_for_block_root(store: Store, block_root: Root) -> Gwei:
+    assert block_root in store.block_states
+
+    # Alternatively, fetch the state from the block's checkpoint
+    state = store.block_states[block_root]
+
+    return get_total_active_balance(state)
+```
+
+#### `get_remaining_weight_in_current_epoch`
+
+```python
+def get_remaining_weight_in_current_epoch(store: Store, block_root: Root) -> Gwei:
+    """ Returns the total weight of votes for this epoch from future committees after the current slot
+    """
+    assert block_root in store.block_states
+
+    # Alternatively, fetches the state from the block's checkpoint
+    state = store.block_states[block_root]
+
+    current_slot = get_current_slot(store)
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(state, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
+```
+
+#### `get_leaf_block_roots`
+
+```python
+def get_leaf_block_roots(store: Store, block_root: Root) -> Set[Root]:
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
```

Comment:
This is very pythonic with the starred expression -- almost anyone reading this spec is going to have to look it up (I did).

Might want to do this with a for loop and each iteration being an explicit union with the last

Reply:
Done.